### PR TITLE
[Repo/Pop] Tensor_RepoPop Draft @ open sesame 11/27 08:31

### PR DESCRIPTION
**Pull request by jijoongmoon**
# PR Description
This is the the another part to complete the tensor
repo. tensor_repopop get the tensor repo data.

**Self evaluation:**
1. Build test:	 [X]Passed [ ]Failed [ ]Skipped
2. Run test:	 [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: jijoong.moon <jijoong.moon@samsung.com>

### CODE REVIEW

File: gst/tensor_reposink/tensor_reposink.c [hunk 1]
```diff
@@ -402,17 +389,17 @@ gst_tensor_reposink_set_caps (GstBaseSink * sink, GstCaps * caps)
   self = GST_TENSOR_REPOSINK (sink);
   gst_caps_replace (&self->in_caps, caps);
 
-  if (DBG) {
+  if (self->silent) {
```
Comment (jaeyun-jung) [inline gst/tensor_reposink/tensor_reposink.c]: change condition !self->silent

File: gst/nnstreamer/tensor_repo.c [hunk 2]
```diff
@@ -27,45 +27,54 @@
...
 {
   gboolean ret = FALSE;
-
-  if (!_repo.initialized)
-    gst_tensor_repo_init ();
+  GstTensorRepoData *new;
 
   GST_REPO_LOCK ();
-  ret = g_hash_table_insert (_repo.hash, GINT_TO_POINTER (myid), data);
+  gpointer *check = g_hash_table_lookup (_repo.hash, GINT_TO_POINTER (nth));
```
[the remark this answers is not in the document]
Comment (jijoongmoon) [reply gst/nnstreamer/tensor_repo.c]: done.

File: gst/nnstreamer/tensor_repo.c [hunk 3]
```diff
@@ -76,44 +85,96 @@ gst_tensor_repo_add_data (GstTensorData * data, guint myid)
...
- * @brief pop GstTensorData from repo
+ * @brief check eos of slot
+ */
+gboolean
+gst_tensor_repo_check_eos (guint nth)
+{
+  GstTensorRepoData *data = gst_tensor_repo_get_repodata (nth);
+  if (data == NULL || data->eos) {
+    if (DBG)
+      g_message ("check eos done [%s]\n", data->eos ? "TRUE" : "FALSE");
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_repo.c]: I'd recommend to use GST_ERROR in this case.

File: gst/nnstreamer/tensor_repo.c [hunk 4]
```diff
@@ -76,44 +85,96 @@ gst_tensor_repo_add_data (GstTensorData * data, guint myid)
...
 
-  GstTensorData *data = gst_tensor_repo_get_tensor (nth);
+  GstTensorRepoData *data = gst_tensor_repo_get_repodata (nth);
   g_return_val_if_fail (data != NULL, FALSE);
 
   data->buffer = buffer;
-  GST_TENSOR_REPO_BROADCAST (nth);
+  if (DBG) {
+    unsigned long size = gst_buffer_get_size (data->buffer);
+    g_message ("Pushed [%d] (size : %lu)\n", nth, size);
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_repo.c]: use GST_DEBUG/GST_INFO/... instead of g_message?

File: gst/tensor_reposink/tensor_reposink.c [hunk 5]
```diff
@@ -186,6 +181,7 @@ gst_tensor_repopush_set_property (GObject * object, guint prop_id,
       break;
     case PROP_SLOT:
       self->myid = g_value_get_uint (value);
+      gst_tensor_repo_add_repodata (self->myid);
```
Comment (myungjoo) [inline gst/tensor_reposink/tensor_reposink.c]: Be aware of dynamism.
PROP_SLOT may get different values in run-time. (changed by user / GUI tools)

File: gst/tensor_reposrc/tensor_reposrc.h [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+/**
+ * @brief GstTensorRepoSrc data structure.
+ *
+ * GstTensorRepoSrc inherits GstBaseSink.
+ */
+struct _GstTensorRepoSrc
+{
+  GstPushSrc parent;
+  GstBaseSink element;
```
Comment (jaeyun-jung) [inline gst/tensor_reposrc/tensor_reposrc.h]: GstBaseSink element;
this line is unnecessary.

File: gst/nnstreamer/tensor_repo.c (hunk 3 above)
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_repo.c]: This line may take a coverity issue (if data is NULL, makes exception 'data->eos')

File: gst/nnstreamer/tensor_repo.c [hunk 7]
```diff
@@ -27,45 +27,54 @@
...
+ * @brief getter to get nth GstTensorRepoData
  */
-GstTensorData *
-gst_tensor_repo_get_tensor (guint nth)
+GstTensorRepoData *
+gst_tensor_repo_get_repodata (guint nth)
 {
-  GstTensorData *data;
+  GstTensorRepoData *data;
   gpointer *p = g_hash_table_lookup (_repo.hash, GINT_TO_POINTER (nth));
```
[the remark this answers is not in the document]
Comment (jijoongmoon) [reply gst/nnstreamer/tensor_repo.c]: Right. Thanks.

File: gst/nnstreamer/tensor_repo.c [hunk 8]
```diff
@@ -27,45 +27,54 @@
...
-  data = (GstTensorData *) p;
+  data = (GstTensorRepoData *) p;
   g_return_val_if_fail (data != NULL, NULL);
```
[the remark this answers is not in the document]
Comment (jijoongmoon) [reply gst/nnstreamer/tensor_repo.c]: Right. I forgot to delete it. Thanks.

File: gst/nnstreamer/tensor_repo.c (hunk 4 above)
Comment (jijoongmoon) [reply gst/nnstreamer/tensor_repo.c]: Right. I will.

File: gst/nnstreamer/tensor_repo.c (hunk 3 above)
Comment (jijoongmoon) [reply gst/nnstreamer/tensor_repo.c]: This is not the error. I think GST_INFO is better. Thanks.

Comment (jijoongmoon) [reply gst/nnstreamer/tensor_repo.c]: I will delete "data == NULL". I will forgot to delete. Thanks.

File: gst/nnstreamer/tensor_repo.c [hunk 9]
```diff
@@ -76,44 +85,96 @@ gst_tensor_repo_add_data (GstTensorData * data, guint myid)
...
+ */
+gboolean
+gst_tensor_repo_check_eos (guint nth)
+{
+  GstTensorRepoData *data = gst_tensor_repo_get_repodata (nth);
+  if (data == NULL || data->eos) {
+    if (DBG)
+      g_message ("check eos done [%s]\n", data->eos ? "TRUE" : "FALSE");
+  }
+  return data->eos;
```
[the remark this answers is not in the document]
Comment (jijoongmoon) [reply gst/nnstreamer/tensor_repo.c]: I will delete "data == NULL". I will forgot to delete. Thanks.

File: gst/tensor_reposink/tensor_reposink.c (hunk 5 above)
Comment (jijoongmoon) [reply gst/tensor_reposink/tensor_reposink.c]: I think it is possible with proper handling the memory slot and lock/cond in tensor repository. Next time we could do to support this dynamism.

File: gst/tensor_reposrc/tensor_reposrc.h (hunk 6 above)
Comment (jijoongmoon) [reply gst/tensor_reposrc/tensor_reposrc.h]: Right. I forgot to delete. Thanks.

File: gst/nnstreamer/tensor_repo.c [hunk 10]
```diff
@@ -27,45 +27,53 @@
...
+  new->buffer = NULL;
+  g_cond_init (&new->cond);
+  g_mutex_init (&new->lock);
+
+  ret = g_hash_table_insert (_repo.hash, GINT_TO_POINTER (nth), new);
   g_assert (ret);
 
-  _print_log ("Successfully added in hash table with key[%d]", myid);
+  if (DBG)
+    g_message ("Successfully added in hash table with key[%d]", nth);
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_repo.c]: Please try "grep g_message ..." on your tensor_repo* directories.

File: gst/tensor_reposrc/tensor_reposrc.c [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+    return GST_FLOW_EOS;
+  }
+
+  if (!self->ini) {
+    int i;
+    guint num_tensors = self->config.info.num_tensors;
+    gsize size = 0;
+    for (i = 0; i < num_tensors; i++)
+      size += gst_tensor_info_get_size (&self->config.info.info[i]);
+    buf = gst_buffer_new_and_alloc (size);
```
[the remark this answers is not in the document]
Comment (jijoongmoon) [reply gst/tensor_reposrc/tensor_reposrc.c]: Done

File: gst/nnstreamer/tensor_repo.c (hunk 10 above)
Comment (jijoongmoon) [reply gst/nnstreamer/tensor_repo.c]: Done. Thanks.

**Pull request merged**